mention: Skip fetching group membership for non-permitted mention.

abdelrahman725 opened the pull request:
Fixes: #32934

Follow up to #33097

Bulk compute groups the sender is allowed to mention, this eliminiates  `O(n)` behaviour in the case where many groups need to be checked for that permission, also we skip fetching membership for those non-allowed-to-mention groups.

Also this handles the edge case where a sender's permission changes while sending the message and after writing `@group_name` in compose box, in this case, we also should NOT fetch membership for that group.

## Changes

1. Add `bulk_get_groups_sender_can_menion` which efficiently computes and fetches groups the sender is allowed to mention, sets the result as `allowed_mention_group_ids` inside `MentionData`, this new bulk function mimics the same logic with the same order of checks of the removed function `sender_can_mention_group`

1. `allowed_mention_group_ids` is then used to check mentioned_groups against, in `init_user_group_data` and in `check_user_group_mention_allowed`.

3. Remove `NamedUserGroup.objects.filter...`, and the `can_mention_group` check for every group, inside `check_user_group_mention_allowed`.

4. Avoid converting `mentioned_group_ids` into a `list`, which saves us an operation.

5. Add test case.

abdelrahman725 commented:
# Important Note

There is a rare edge cases which results in `O(n)` queries for the groups mentioned: if the user attempts to mention multiple groups for all of which they don't have permission to mention.

This is still avoidable, but needs more work, Is that edge case worth optimizing ?

abdelrahman725 commented on zerver/lib/message.py:
```diff
@@ -1507,16 +1506,14 @@ def stream_wildcard_mention_allowed(sender: UserProfile, stream: Stream, realm:
+    for group_id in mentioned_group_ids:
+        if group_id not in mention_data.allowed_mention_group_ids:
             raise JsonableError(
                 _("You are not allowed to mention user group '{user_group_name}'.").format(
-                    user_group_name=group.name
+                    user_group_name=mention_data.get_user_group_name(group_id)
                 )
             )
 
```
Should this nclude all the non-allowed gorups ? instead of just raising `JsonableError` for the first found non-allowed group

abdelrahman725 commented on zerver/lib/message.py:
```diff
@@ -1507,16 +1506,14 @@ def stream_wildcard_mention_allowed(sender: UserProfile, stream: Stream, realm:
     return can_mention_many_users(sender)
 
 
-def check_user_group_mention_allowed(sender: UserProfile, user_group_ids: list[int]) -> None:
```
I safely removed `sender` parameters, as `mention_data` is already tied to single message for a single user.

abdelrahman725 commented:
# Updates

Create `bulk_get_groups_sender_can_menion` which fetches groups the sender can mention in a single efficient query,

After this PR, We can happily say that all `O(n)` behaviour associated with mentioning multiple groups in a single message is avoided for all use-cases !

alya commented:
@timabbott Are you reviewing this one directly?

abdelrahman725 commented on zerver/lib/mention.py:
```diff
@@ -464,23 +476,58 @@ def get_user_group_mention_display_name(user_group: NamedUserGroup) -> StrPromis
 
-def sender_can_mention_group(sender: UserProfile | None, named_group: NamedUserGroup) -> bool:
-    can_mention_group = named_group.can_mention_group
+def bulk_get_groups_sender_can_menion(
```
ok

abdelrahman725 commented:
# updates

1. Avoid fetching all members of each mentioned group, and instead only fetch the set of IDs of all direct/indirect groups the sender is member, then check if `can_mention_group` of each mentioned_group is in that fetched set.

2.  Add 2 test functions to ensure :

- We correctly do the mention permission check when permission changes.
- We do it in a constant time for the case  many groups (which need permission check) are mentioned in a single message.

timabbott commented:
I think the next step for this PR is maintainer review with @shubham-padia, since he's quite familiar with this class of performance reasoning at this point.

shubham-padia commented on zerver/lib/mention.py:
```diff
@@ -393,37 +393,47 @@ def message_has_topic_wildcards(self) -> bool:
+            for group in named_user_groups:
+                group_name = group.name.lower()
+                self.user_group_name_info[group_name] = group
+                self.user_group_names[group.id] = group_name
```
Since we don't need `.user_group_names`, we can go back to using the same code as before:

```
self.user_group_name_info = {group.name.lower(): group for group in named_user_groups}
```

shubham-padia replied:
Yeah, you're right. We're storiing the lower case group_name in user_group_names and we use that in `get_user_group_name`, which is used in the following code block

```
raise JsonableError(
                _("You are not allowed to mention user group '{user_group_name}'.").format(
                    user_group_name=mention_data.get_user_group_name(group_id)
                )
            )
```

We should be having the non lower-cased group name here.

shubham-padia replied:
> We should be having the non lower-cased group name here.

Has this been addressed?

abdelrahman725 replied:
Yea, sorry. Addressed now.

shubham-padia commented on zerver/lib/mention.py:
```diff
@@ -393,37 +393,47 @@ def message_has_topic_wildcards(self) -> bool:
+            )
 
-            # We only fetch group membership mentions that can
+            # We only fetch group membership that can
```
`group membership` -> `group memberships`

shubham-padia commented on zerver/tests/test_markdown.py:
```diff
@@ -395,6 +396,144 @@ def test_silent_mention_user_groups(self) -> None:
+        # that extra query is only triggered when any mentioned group
+        # needs permission check (e.g. not SystemGroups.EVERYONE).
+        with self.assert_database_query_count(3):
+            mention_data = MentionData(mention_backend, content, message_sender=zoe)
```
We should also have one test for checking query count when every mentioned group has can_mention_group as EVERYONE

shubham-padia reviewed (commented):
This is great! Did a first pass.

abdelrahman725 commented on zerver/lib/mention.py:
```diff
@@ -450,6 +460,9 @@ def get_user_ids(self) -> set[int]:
         return self.user_group_name_info.get(name.lower(), None)
 
+    def get_user_group_name(self, group_id: int) -> str:
+        return self.user_group_names.get(group_id, "")
```
How ?

I use it in [check_user_group_mention_allowed](https://github.com/zulip/zulip/pull/34484/files#diff-40c1c91d1032197fa2decf32d350e72180d5861534c2eb99fcc898eb0b12da24R1535), to get the group name by id, so it can be passed in the json response

abdelrahman725 committed:
mention: Skip fetching group membership for non-permitted mention.

Fixes: #32934

Follow up to #33097

Bulk compute groups the senders is allowed to mention, this eliminiates
the O(n) behaviour in case many groups needs to be checked
for mention permission, also we skip fetching membership for those
non-allowed-to-mention groups.

1. Add bulk_get_groups_sender_can_mention
which efficiently computes and fetches groups the sender
is allowed to mention and sets the
result as `allowed_mention_group_ids`
inside `MentionData`

1. `allowed_mention_group_ids` is then
used to check mentioned_groups against,
in `init_user_group_data` and in
`check_user_group_mention_allowed`.

2. Remove `NamedUserGroup.objects.filter...`,
and the `can_mention_group` check for every group,
inside `check_user_group_mention_allowed`.

3. Avoid converting `mentioned_group_ids`
into a `list`, which saves us an operation.

4. Add 2 test functions to ensure we correctly do
the mention permission check when permission changes
and we do it in a constant time for the case many groups

shubham-padia commented on zerver/lib/mention.py:
```diff
@@ -482,23 +495,51 @@ def get_user_group_mention_display_name(user_group: NamedUserGroup) -> StrPromis
+        ):
+            group_ids_everyone_can_mention.add(named_group.id)
+        else:
+            group_ids_requiring_permission_check.add(named_group)
```
group_ids_requiring_permission_check does not contain group ids, but rather groups, probably should rename it to `groups_requiring_permission_check`

shubham-padia commented on zerver/lib/mention.py:
```diff
@@ -482,23 +495,51 @@ def get_user_group_mention_display_name(user_group: NamedUserGroup) -> StrPromis
-        direct_member_only=False,
+    setting_config = NamedUserGroup.GROUP_PERMISSION_SETTINGS["can_mention_group"]
+
+    assert sender is not None
```
For channel/channel folder descriptions, sender will be none. If we have a group with can_mention_group set to something other than everyone, we will get a 500 since it is not a message and sender is none. 

For this PR, we should restore the previous behaviour of allowing these mentions for such cases. I'm not sure how tricky that might be.

We should also start a separate discussion on CZO about whether we should be allowing mentions from users who cannot mention a group in a message but can mention in a channel description. It might be the case that the previous behaviour was not a conscious decision and just a side effect of how code was written.

abdelrahman725 replied:
Nice catch! 

**main** doesn't actually check for permission in case of a group mention inside a channel description: `sender_can_mention_group`  (precedent of `bulk_get_groups_sender_can_mention`) is not called. 

Started [#backend > Handle Group mention permission](https://chat.zulip.org/#narrow/channel/3-backend/topic/Handle.20Group.20mention.20permission/with/2507236)

shubham-padia replied:
Great! 

> For this PR, we should restore the previous behaviour of allowing these mentions for such cases. I'm not sure how tricky that might be.

How much work would this be? Ideally we would preserve the current behaviour and then address that thread in a different PR

abdelrahman725 replied:
It's trivial to preserve main's behavior: when `message_sender` is `None` inside  `MentionData.init_user_group_data`, we don't check for permission.

But that doesn't necessarily mean it's the right approach, but yeah in a different PR is fine

abdelrahman725 replied:
**Updated**

We now skip permission check when sender is None as main does.

abdelrahman725 replied:
Opened #39962

shubham-padia reviewed (commented):
Left some comments.

abdelrahman725 commented on zerver/lib/mention.py:
```diff
@@ -397,37 +397,55 @@ def message_has_topic_wildcards(self) -> bool:
                 realm_for_sharding_id=realm_id, name__in=user_group_names_mentions
-            )
-
-            # No filter here as we need user_group_name_info for all groups mentions.
-            self.user_group_name_info = {group.name.lower(): group for group in named_user_groups}
+            ).select_related("can_mention_group", "can_mention_group__named_user_group")
```
Yes, `.select_related(...)` will be redundant when `self.message_sender` is None, but I believe that's fine: It's not that common and to not complicate the code.

abdelrahman725 commented on zerver/lib/mention.py:
Constructing `self.allowed_mention_group_ids` is actually redundant and can be avoidable when `self.message_sender` is None: 
We simply don't match `group_ids_to_fetch_membership_for` against  `self.allowed_mention_group_ids` . 

But for consistency and simplicity, I kept it this way.

shubham-padia commented on zerver/lib/mention.py:
```diff
@@ -397,37 +397,55 @@ def message_has_topic_wildcards(self) -> bool:
+                self.user_group_name_info[group_name] = group
+                self.user_group_names[group.id] = group_name
+
+            # message_sender cane be None when mentioning
```
cane -> can

shubham-padia commented on zerver/lib/mention.py:
```diff
@@ -486,23 +507,52 @@ def get_user_group_mention_display_name(user_group: NamedUserGroup) -> StrPromis
+def bulk_get_groups_sender_can_mention(
+    sender: UserProfile, named_user_groups: QuerySet[NamedUserGroup]
+) -> set[int]:
 
```
We usually don't have this blank line between the function and the first statement.

shubham-padia reviewed (commented):
Left some minor comments, in general looks good.

shubham-padia commented:
Looks good to go to integration review, tagging @sahil839 since this is groups related!

sahil839 commented on zerver/lib/mention.py:
```diff
@@ -486,23 +506,51 @@ def get_user_group_mention_display_name(user_group: NamedUserGroup) -> StrPromis
-        NamedUserGroup.GROUP_PERMISSION_SETTINGS["can_mention_group"],
-        direct_member_only=False,
+    if is_cross_realm_bot_email(sender.delivery_email) or (
+        not setting_config.allow_everyone_group and sender.is_guest
```
`allow_everyone_group` is `True` for this setting so this is redundant. If we want to keep it as a safeguard for future then maybe we can also add a check for the group being set to `EVERYONE_ON_INTERNET` even though we cannot set any setting to that group currently.

sahil839 commented on zerver/lib/mention.py:
```diff
@@ -486,23 +506,51 @@ def get_user_group_mention_display_name(user_group: NamedUserGroup) -> StrPromis
+
+    # Ids of all direct and indirect groups the sender is a member of.
+    user_recursive_group_ids = set(
+        get_recursive_membership_groups(sender).values_list("id", flat=True)
```
We fetch recursive groups already for checking other group permissions, so can that be passed to `MentionData`?

sahil839 commented on zerver/tests/test_markdown.py:
```diff
@@ -398,6 +381,155 @@ def test_silent_mention_user_groups(self) -> None:
+        # This also tests an edge case where a sender's permission changes
+        # (i.e. they can't mention that group) while sending the message
+        # and after writing @group_name in compose box,
+        # in which case we should NOT fetch membership.
```
How does this check the case when permission changes while sending the message and after writing group name in compose box?

sahil839 commented on zerver/tests/test_markdown.py:
```diff
@@ -398,6 +381,155 @@ def test_silent_mention_user_groups(self) -> None:
+        test_group = check_add_user_group(realm, "test_group", [aaron], acting_user=iago)
+
+        sub_group_lvl_1 = check_add_user_group(realm, "sub_group_lvl_1", [zoe], acting_user=iago)
+        sub_group_lvl_2 = check_add_user_group(
```
I think naming the groups here as `aaron_group`, `zoe_group` and `prospero_group` would be more readable.
